src/query/plan/operator.cpp

```diff
@@ -4024,4 +4026,76 @@ UniqueCursorPtr LoadCsv::MakeCursor(utils::MemoryResource *mem) const {
   return MakeUniqueCursorPtr<LoadCsvCursor>(mem, this, mem);
 };
 
+class ForeachCursor : public Cursor {
+ public:
+  explicit ForeachCursor(Symbol output_symbol, UniqueCursorPtr input, Expression *expr, utils::MemoryResource *mem,
+                         bool is_nested)
+      : output_symbol_(output_symbol),
+        input_cursor_(std::move(input)),
+        expression(expr),
+        cache_(mem),
+        is_nested_(is_nested) {}
+
+  bool Pull(Frame &frame, ExecutionContext &context) override {
+    SCOPED_PROFILE_OP(op_name_);
+
+    if (cache_.empty()) {
+      if (!input_cursor_->Pull(frame, context)) {
+        return false;
+      }
+
+      ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                    storage::View::NEW);
+      TypedValue expr_result = expression->Accept(evaluator);
+      if (!expr_result.IsList()) {
+        throw QueryRuntimeException("FOREACH expression must resolve to a list, but got '{}'.", expr_result.type());
+      }
+      cache_ = expr_result.ValueList();
+      index_ = cache_.begin();
+    }
+
+    if (index_ == cache_.end()) {
+      if (!is_nested_ || !input_cursor_->Pull(frame, context)) {
+        return false;
+      }
+      index_ = cache_.begin();
+    }
```

Comment:
I think checking two different things for pulling is not good. I think `index_` should be initialized in the ctor with `cache_.end()`, then `index_ == cache_.end()` can be used to check whether another pull is necessary or not.
```suggestion
    while (index_ == cache_.end()) {
      if (!input_cursor_->Pull(frame, context)) {
        return false;
      }

      ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
                                    storage::View::NEW);
      TypedValue expr_result = expression->Accept(evaluator);
      if (!expr_result.IsList()) {
        throw QueryRuntimeException("FOREACH expression must resolve to a list, but got '{}'.", expr_result.type());
      }
      cache_ = expr_result.ValueList();
      index_ = cache_.begin();
    }
```

Furthermore I don't see why it is important to know whether we are in a nested foreach clause or not. How does that should alter the behavior?

Reply:
Some queries that I think point out the problems:
```
memgraph> MATCH (n) DETACH DELETE n;
Empty set (0.001 sec)
memgraph> CREATE (:V {prop: [[1, 2], []]});
Empty set (0.000 sec)
memgraph> MATCH (n) RETURN n;
+---------------------------+
| n                         |
+---------------------------+
| (:V {prop: [[1, 2], []]}) |
+---------------------------+
1 row in set (0.027 sec)
memgraph> MATCH (n) FOREACH ( li IN n.prop | FOREACH ( value IN li | CREATE (m:F {prop:value})));
Empty set (0.001 sec)
memgraph> MATCH (n) RETURN n;
+---------------------------+
| n                         |
+---------------------------+
| (:V {prop: [[1, 2], []]}) |
| (:F {prop: 1})            |
| (:F {prop: 2})            |
| (:F {prop: 1})            |
| (:F {prop: 2})            |
+---------------------------+
5 rows in set (0.001 sec)
```
If the original node is created as `CREATE (:V {prop: [[], [1, 2]]});`, then memgraph crashes.

In my opinion rewriting the cursor with the proposed changes fixes this issue.